### PR TITLE
AMQP 1.0 binary parser: treat arrays with extra or missing input as fatal errors

### PR DESCRIPTION
**Pull request by michaelklishin**
## Proposed Changes

With some input it is possible that the terminating clause will never
match.

While at it, consume binary input when parsing short form
primitives: null, true, false, as well as uint/ulong zero values.

## Types of Changes

- [x] Bug fix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)

## Checklist

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in related repositories

## Further Comments

Pair: @lhoguin.

Per discussion with @kjnilsson.

### CODE REVIEW

**Review by lhoguin — APPROVED** [review]
Approved based on re-reading of the changes. I have no tested them on my machine.

File: deps/amqp10_common/src/amqp10_binary_parser.erl [hunk 1]
```diff
@@ -122,6 +122,14 @@ parse_compound1(Count, Bin, Acc) ->
     {Value, Rest} = parse(Bin),
     parse_compound1(Count - 1, Rest, [Value | Acc]).
 
+parse_array_primitive(16#40, <<_:8/unsigned, R/binary>>) -> {null, R};
+parse_array_primitive(16#41, <<_:8/unsigned, R/binary>>) -> {true, R};
+parse_array_primitive(16#42, <<_:8/unsigned, R/binary>>) -> {false, R};
+parse_array_primitive(16#43, <<_:8/unsigned, R/binary>>) -> {{uint, 0}, R};
+parse_array_primitive(16#44, <<_:8/unsigned, R/binary>>) -> {{ulong, 0}, R};
+parse_array_primitive(Marker, Data) ->
```
Comment (kjnilsson) [inline deps/amqp10_common/src/amqp10_binary_parser.erl:130]: Maybe use 'Type' here instead of 'Marker' for consistency?

**Review by kjnilsson — APPROVED** [review]
Just one minor naming comment

File: deps/amqp10_common/src/amqp10_binary_parser.erl (hunk 1 above)
Comment (michaelklishin) [reply deps/amqp10_common/src/amqp10_binary_parser.erl]: Done

**Comment by lhoguin** [comment]
Reproduced the issue before the patch and confirmed that the patch fixes it.

**Comment by michaelklishin** [comment]
Backported to `v3.8.x`.

**Pull request merged**